packages/core/src/file/file.element.ts

```diff
@@ -49,36 +49,53 @@ export class CdsFile extends CdsControl {
           <cds-icon shape="folder" aria-hidden="true"></cds-icon>
           ${this.buttonLabel}
         </cds-button>
-        ${this.clearFiles}
+        ${this.clearFilesControlTemplate}
       </div>
     `;
   }
 
-  protected get clearFiles() {
-    return this.inputControl.files?.length
-      ? html` <cds-control-action @click="${() => this.updateLabelAndFocus()}" aria-label="${this.i18n.removeFile}">
+  protected get clearFilesControlTemplate() {
+    return this.inputControl.files?.length && !this.disabled
+      ? html` <cds-control-action @click="${() => this.clearFiles()}" aria-label="${this.i18n.removeFile}">
           <cds-icon shape="times"></cds-icon>
         </cds-control-action>`
       : html``;
   }
 
   firstUpdated(props: Map<string, any>) {
     super.firstUpdated(props);
-    this.inputControl.addEventListener('change', e => this.updateLabelAndFocus((e.target as any).files));
+    (this.inputControl as Element).addEventListener('change', e => {
+      // NOTE: have to distinguish here being user-caused events and programmatic
+      // events (e.isTrusted true/false) so that we don't fire a change event loop
+      if (e.isTrusted) {
+        this.updateLabelAndFocus((e.target as any).files);
+      }
+    });
+  }
+
+  /** @private */
+  clearFiles(fireEvent = true) {
+    this.buttonLabel = this.i18n.browse;
+    this.inputControl.value = '';
+
+    // when input is reset like this it isn't registering an onchange event
+    // NOTE: tsc + karma is complaining about dispatchEvent being a method on inputControl
+    if (fireEvent && this.inputControl.dispatchEvent) {
+      (this.inputControl as Element).dispatchEvent(new Event('change'));
+    }
+
+    const browseButton = this.shadowRoot?.querySelector('cds-button');
+    if (browseButton) {
+      browseButton.focus();
+    }
   }
 
-  private updateLabelAndFocus(files?: FileList) {
+  /** @private */
```

Comment:
Do we go in this direction here just to be able to unit test this function?

Reply:
Yep. I improved code coverage in this file from 80% to way over 90% and this helped me do that. `@private` makes sure this doesn't show up in our docs. Almost none of this method was covered by unit tests before.